### PR TITLE
common/pick_address: filter out loopback addresses

### PR DESCRIPTION
**Pull request by tchaikov**
instead of filtering out loopback ifaces, check for loopback addresses.

before this change, iface named "lo" is filtered out by default,
and "lo" is allowed if `ms_bind_exclude_lo_iface` is false.

after this change, iface with loopback address of "127.0.0.1" is
filtered out. also, the iface marked down is not considered.

the option of "ms_bind_exclude_lo_iface" is removed. the tests are
updated accordingly.

Fixes: https://tracker.ceph.com/issues/50456
Signed-off-by: Kefu Chai <kchai@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by dvanders** [comment]
Do we need to filter out all of 127/8 or just INADDR_LOOPBACK?
Anyway, this approach seems usable, but I would've thought that having a config `ms_bind_exclude_lo_iface` to optionally allow binding to `lo` (in those cases like zigo's) was more flexible. Do we support single-box ceph clusters with only local network, for testing at least? This would kill it, right?

I also have another observation in this area that I wanted to clear up. Empirically, what we currently do (in nautilus at least, OOTB, with default addr/network config) is pick a non-local address to register in the osdmap (so other OSDs can find this one); *but* when it comes to binding to an interface/addr, it seems to me that we bind to 0.0.0.0:<port>. For example, this osd is listening on all addrs:
```
tcp        0      0 0.0.0.0:6980            0.0.0.0:*               LISTEN      870636/ceph-osd
```
but the osdmap knows where he is:
```
osd.35 up   in  weight 1 up_from 367490 up_thru 367490 down_at 367482 last_clean_interval [365212,367478) [v2:128.142.162.21:6968/870636,v1:128.142.162.21:6973/870636] [v2:128.142.162.21:6977/870636,v1:128.142.162.21:6980/870636] exists,up 7c6bf9f1-c527-421c-8237-029d53c2cbd2
```
And by observation if we configure public/cluster addr to 128.142.162.21, then we'll get the same result in the osdmap, but the daemon now LISTENs on 128.142.162.21 only.

Is this an unimportant side-effect or was listening on 0.0.0.0 intentional? I don't really know what is best practice here, but I would have assumed it is best to always listen on 0.0.0.0 for both cluster/public addrs, while advertising the real IPs in the osdmap.

**Comment by tchaikov** [comment]
@thomasgoirand  hi Thomas, could you help to double check if this is what you are after? i am find either way.

**Comment by tchaikov** [comment]
@dvanders hi Dan, thanks for your insights. 

> Do we need to filter out all of 127/8 or just INADDR_LOOPBACK?

i am not sure.

> but I would've thought that having a config `ms_bind_exclude_lo_iface` to optionally allow binding to `lo` (in those cases like zigo's) was more flexible. 

i thought `INADDR_LOOPBACK` was what zigo asked for. let's wait for his input. once we are the same page on

- adding `ms_bind_exclude_lo_iface` option or not
- check "lo" and or and/or `IFF_LOOPBACK` flags or
- check address of 127/8 or `INADDR_LOOPBACK`.

i will update this changeset.

**Comment by thomasgoirand** [comment]
Hi,

What I am really doing is exactly that: put the IP address of the machine on the lo interface. Then after, frr (the routing daemon) really is advertizing what's on the local loopback. To give a bit of insight, here's my configuration.

In our setup, we have 2x 25 GBits/s port. Each of them catches a IPv6 link local address, which gives it local connectivity to the 2 neighbor switches. The machine then uses this IPv6 connectivity to transport IP packets using BGP (with the FRR daemon). Every IP which is added to the lo interface is advertized by FRR.

Why are we doing this way? Because we got tired of huge ARP spanning trees that get reset each time a switch is added. Because that way, we don't have any broadcast traffic. The only thing we get is IPv6 link local auto-discovery, and that is limited between the switches and the host itself. It also is very fault tolerant, and routing can be done in all directions using multiple equipment (ECMP routing). That's much nicer than bonding and L2.

So yeah, I do want ceph to bind on 10.1.1.6, which is on the lo interface. Please don't break that setup.

Here are the config files below (Debian style, but it should be working quite the same way on RPM based systems). You'll notice that on top of my above explanations, we're using a VLAN for the BGP traffic, but that doesn't change anything of the explanations (you can ignore that part if you like). The metric 4278198272 is to tell Linux not to use the interface directly, and use BGP for routing packets through the VLAN (so we give a huge metric so the route isn't used).

# cat /etc/network/interfaces
# This file describes the network interfaces available on your system
# and how to activate them. For more information, see interfaces(5).

# The loopback network interface
auto lo
iface lo inet loopback

auto eth0
iface eth0 inet manual
	metric 4278198272
	mtu 9000

auto eth0.10
iface eth0.10 inet manual
	mtu 9000

auto eth1
iface eth1 inet manual
	metric 4278198272
	mtu 9000

auto eth1.10
iface eth1.10 inet manual
	mtu 9000


# cat /etc/frr/frr.conf 
# default to using syslog. /etc/rsyslog.d/45-frr.conf places the log
# in /var/log/frr/frr.log
log syslog informational
log syslog informational

!
service integrated-vtysh-config
!
log syslog informational
!
int lo
 ip address 10.1.1.6/32
!
router bgp <THE-AS-OF-MY-CEPH-CLUSTER>
 bgp router-id 10.1.1.6
 bgp bestpath as-path multipath-relax
 bgp bestpath compare-routerid
 neighbor pg-leaf peer-group
 neighbor pg-leaf remote-as external
 neighbor pg-leaf capability extended-nexthop
 neighbor eth0.10 interface peer-group pg-leaf
 neighbor eth1.10 interface peer-group pg-leaf
 !
 address-family ipv4 unicast
  redistribute connected route-map map-redistribute
  neighbor pg-leaf soft-reconfiguration inbound
  neighbor pg-leaf route-map map-leaf-in in
  neighbor pg-leaf route-map map-leaf-out out
 exit-address-family
 !
 address-family ipv6 unicast
  redistribute connected route-map map-redistribute
  neighbor pg-leaf activate
  neighbor pg-leaf soft-reconfiguration inbound
  neighbor pg-leaf route-map map-leaf-in in
  neighbor pg-leaf route-map map-leaf-out out
 exit-address-family
!
route-map map-redistribute permit 10
 match interface lo
!
route-map map-leaf-out permit 10
 match interface lo
!
route-map map-leaf-in permit 10
!

**Comment by tchaikov** [comment]
@dvanders @thomasgoirand per @dvanders 's great suggestion, i revise this change, so ceph prefers the non-loopback address and up iface when picking iface and addresses, but it does **not** skip loopback address of loopback iface anymore.

File: src/test/test_ipaddr.cc [hunk 1]
```diff
@@ -205,18 +207,13 @@ TEST(CommonIPAddr, TestV4_SkipLoopback)
   result =
     find_ip_in_subnet_list(nullptr, (struct ifaddrs*)&one,
                            CEPH_PICK_ADDRESS_IPV4 | CEPH_PICK_ADDRESS_IPV6,
-                           "10.0.0.0/8", "", true);
+                           "10.0.0.0/8", "");
```
Comment (dvanders) [inline src/test/test_ipaddr.cc:210]: We should remove the 10.0.0.0/8 subnet filter here to verify that we correctly prefer the 10.1.2.3 despite the two loopback addresses.

File: src/test/test_ipaddr.cc [hunk 2]
```diff
@@ -341,18 +339,13 @@ TEST(CommonIPAddr, TestV6_SkipLoopback)
   result =
     find_ip_in_subnet_list(nullptr, (struct ifaddrs*)&one,
                            CEPH_PICK_ADDRESS_IPV4 | CEPH_PICK_ADDRESS_IPV6,
-                           "2001:1234:5678:90ab::0/64", "", true);
+                           "2001:1234:5678:90ab::0/64", "");
```
Comment (dvanders) [inline src/test/test_ipaddr.cc:342]: Same here. We need these tests to confirm that non-loopback are not chose even when we don't specify a subnet.

Comment (tchaikov) [reply src/test/test_ipaddr.cc]: @dvanders thank you for your review. fixed and repushed. could you please take another look?

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-04-23_14:12:53-rados-wip-kefu-testing-2021-04-23-2026-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-24_04:07:09-rados-wip-kefu-testing-2021-04-23-2026-distro-basic-smithi/

**Comment by tchaikov** [comment]
@jdurgin @neha-ojha could you take a look? as @thomasgoirand is preparing the debian packaging of nautilus, it'd be great if we can have this change in master sooner, so i can have enough confidence to backport it to nautilus to unblock him.

**Comment by tchaikov** [comment]
@jdurgin @liewegas @neha-ojha ping?

**Review by jdurgin — CHANGES_REQUESTED** [review]
@dvanders I don't think this changes the 0.0.0.0 binding - @tchaikov can you confirm?

The grading scheme is elegant, however I don't think it should change existing clusters' behavior beyond the loopback handling. Other changes could break an existing deployment if the wrong address is chosen. Even if it could be avoided by specifying the desired subnet/interface, we should try to avoid even accidental breakage of this sort.

For example, even though  ```osd_numa_prefer_iface``` has 'prefer' in the name, it's true by default, and the implementation is actually treating it as a requirement. This means many existing clusters may be getting one address today, but could get a different one now, particularly if there's a startup race and the currently used if is down when addresses are picked, but comes up later.

Why not keep it simple and only change the loopback behavior with this patch?

**Comment by tchaikov** [comment]
> @dvanders I don't think this changes the 0.0.0.0 binding - @tchaikov can you confirm?

@jdurgin this change is not related to the 0.0.0.0 binding.

> 
> The grading scheme is elegant, however I don't think it should change existing clusters' behavior beyond the loopback handling. Other changes could break an existing deployment if the wrong address is chosen. Even if it could be avoided by specifying the desired subnet/interface, we should try to avoid even accidental breakage of this sort.
> 
> For example, even though `osd_numa_prefer_iface` has 'prefer' in the name, it's true by default, and the implementation is actually treating it as a requirement. This means many existing clusters may be getting one address today, but could get a different one now, particularly if there's a startup race and the currently used if is down when addresses are picked, but comes up later.
>

please note, the implementation is not treating it as a hard requirement. what it does is to take it into consideration while grading the addresses.

> Why not keep it simple and only change the loopback behavior with this patch?

yes, it's simpler with this change.

**Review by jdurgin — APPROVED** [review]
Apologies, I misread the existing numa code - the new code is clearer :)

The only new thing then is loopback handling and preferring up addrs, which sounds quite low risk to me.

**Comment by tchaikov** [comment]
this change broke the win32 build
```
../src/common/pick_address.cc: In function 'int {anonymous}::is_loopback_addr(sockaddr*)':
../src/common/pick_address.cc:53:11: error: 'in_addr_t' does not name a type; did you mean 'in_addr'?
   53 |     const in_addr_t net = ntohl(sin->sin_addr.s_addr) >> IN_CLASSA_NSHIFT;
      |           ^~~~~~~~~
      |           in_addr
../src/common/pick_address.cc:54:12: error: 'net' was not declared in this scope
   54 |     return net == IN_LOOPBACKNET ? 1 : 0;
      |            ^~~
../src/common/pick_address.cc:54:19: error: 'IN_LOOPBACKNET' was not declared in this scope; did you mean 'IFF_LOOPBACK'?
   54 |     return net == IN_LOOPBACKNET ? 1 : 0;
      |                   ^~~~~~~~~~~~~~
      |                   IFF_LOOPBACK
../src/common/pick_address.cc:52:24: warning: unused variable 'sin' [-Wunused-variable]
   52 |     const sockaddr_in* sin = (struct sockaddr_in *)(addr);
      |                        ^~~
```

**Comment by tchaikov** [comment]
will create a fix next Monday.

**Comment by tchaikov** [comment]
#41116

**Pull request merged**